### PR TITLE
Improved logic of Boyer Moore & added few edge cases in test

### LINKED ISSUE #174: Boore moore algorithm is not working
https://github.com/williamfiset/Algorithms/blob/master/src/main/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearch.java

text = AABAACAADAABAABA
pattern = AABA
not working for this pattern and text.

### PR DESCRIPTION
**Pull request by sharoha**
This PR fixes #174 for edge cases. It uses Bad Character Heuristic to find the matches.

### CODE REVIEW

File: src/main/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearch.java [hunk 1]
```diff
@@ -21,11 +22,7 @@
       return new ArrayList<>();
     }
     List<Integer> occurrences = new ArrayList<>();
-
-    Map<Character, Integer> skipTable = new HashMap<>();
-    for (int i = 0; i < pattern.length(); i++) {
-      skipTable.put(pattern.charAt(i), pattern.length() - i - 1);
-    }
+    Map<Character, TreeSet<Integer>> skipTable = generateSkipTable(pattern);
```
Comment (williamfiset) [inline src/main/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearch.java:25]: I think using a Map + a TreeMap is huge overkill for this algorithm that will cause lots of memory overhead and slowness. I had a look at 3+ implementations, and none of the used a data structure like this.

Please refactor and have a look at some examples:
1. http://www.personal.kent.edu/~rmuhamma/Algorithms/MyAlgorithms/StringMatch/boyerMoore.htm
2. https://www.geeksforgeeks.org/boyer-moore-algorithm-for-pattern-searching/

File: src/test/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearchTest.java [hunk 2]
```diff
@@ -36,11 +36,14 @@ public void shouldReturnOneOccurrence() {
...
   public void shouldReturnMultiplyOccurrences() {
+    assertThat(underTest.findOccurrences("SAAT TE", "TE")).containsExactly(5);
     assertThat(
             underTest.findOccurrences("Sample text for testing the Boyer-Moore algorithm.", "te"))
         .containsExactly(7, 16);
     assertThat(underTest.findOccurrences("Sample text for testing the Boyer-Moore algorithm.", " "))
         .containsExactly(6, 11, 15, 23, 27, 39);
+    assertThat(underTest.findOccurrences("AABAACAADAABAABA", "AABA")).containsExactly(0, 9, 12);
+    assertThat(underTest.findOccurrences("AAAAAAA", "AA")).containsExactly(0, 1, 2, 3, 4, 5);
   }
```
Comment (williamfiset) [inline src/test/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearchTest.java:47]: To make sure the Boyer Moore impl is 100% correct, can also add a test that verifies a long string against a brute force implementation that finds patterns a text?

Comment (sharoha) [reply src/test/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearchTest.java:47]: Yeah sure, this makes sense. I will add a brute force implementation in the test and randomly generate a large string that will validate the testing in a dynamic way.

File: src/test/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearchTest.java [hunk 3]
```diff
@@ -35,12 +42,78 @@ public void shouldReturnOneOccurrence() {
...
+  public void shouldReturnEmptyForPatternLengthLargerThenText() {
+    assertThat(underTest.findOccurrences("This is a Test Text", "This is a test Pattern"))
+        .isEmpty();
+  }
+
+  @Test
+  public void shouldReturnDynamicString() {
+    int runLength = random.nextInt(MAX_ITERATION) + 1;
+    for (int run = 0; run < runLength; run++) {
+      int upperCharText = random.nextInt(26);
```
Comment (williamfiset) [inline src/test/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearchTest.java:66]: Suggestion: To increase the number of matches restrict the alphabet size. For instance, make the text and pattern only from the alphabet {'A' and 'B'} rather than {'A', 'B', ..., 'Z'}

File: src/test/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearchTest.java (hunk 2 above)
Comment (williamfiset) [reply src/test/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearchTest.java:54]: Thanks!

File: src/test/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearchTest.java (hunk 3 above)
Comment (sharoha) [reply src/test/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearchTest.java]: Added code to limit the number of characters.

File: src/main/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearch.java (hunk 1 above)
Comment (williamfiset) [reply src/main/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearch.java]: I think this can still use some improvements. I see a lot of potential to improve the runtime with very simple optimizations. I can handle this myself later, the important thing is that the algorithm is now fixed.

**Comment by williamfiset** [comment]
Rohan, thank you for adding tests and fixing this implementation!

File: src/main/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearch.java (hunk 1 above)
Comment (sharoha) [reply src/main/java/com/williamfiset/algorithms/strings/BoyerMooreStringSearch.java]: Can you point a few of those improvements, curious to know!!!
